pwnlib/libcdb.py

```diff
@@ -100,11 +110,76 @@ def provider_libc_rip(hex_encoded_id, hash_type):
         return None
     return data
 
+# Search offline https://github.com/niklasb/libc-database for symbols
+def find_local_libc_database(params):
+    if not local_database["libc-database"]:
+        log.warn_once("No environment variable LOCAL_LIBC_DATABASE or parameter libcdb.local_database['libc-database'] are set")
+        return None
+    
+    if not params.get("symbols"):
+        return None
+
+    res = []
+    db_path = Path(local_database["libc-database"]) / "db"
+
+    for symbols_path in db_path.rglob("*.symbols"):
+        syms = get_libc_symbols(symbols_path)
+
+        matched = 0
+        for name, addr in params["symbols"].items():
+            if isinstance(addr, str):
+                addr = int(addr, 16) 
+
+            if syms.get(name) and syms.get(name) & 0xfff == addr & 0xfff:
+                matched += 1
+
+        if matched == len(params["symbols"]):
+            res.append(get_libc_info(db_path, symbols_path.stem, syms))
+
+    return res
+
+# Offline search https://github.com/niklasb/libc-database for hash type
+def offline_provider_libc_database(hex_encoded_id, hash_type):
+    if not local_database["libc-database"]:
+        log.warn_once("No environment variable LOCAL_LIBC_DATABASE or parameter libcdb.local_database['libc-database'] are set")
+        return None
+
+    db_path = Path(local_database["libc-database"]) / "db"
+    hashfilehex = None
+
+    if hash_type == "build_id":
+        hash_type = "buildid"
+
+    # "buildid" is not in HASHES, so we can correct calculate file hash
+    if hash_type in HASHES:
+        hashfilehex = eval(f"{hash_type}filehex")
+
+    for libc_path in db_path.rglob("*.so"):
+        if hash_type == "buildid" and hex_encoded_id == get_build_id(libc_path):
+            return read(libc_path)
+        elif hashfilehex is not None and hex_encoded_id == hashfilehex(libc_path):
+            return read(libc_path)
+
+    return None
+
+
 PROVIDERS = [provider_libcdb, provider_libc_rip]
+OFFLINE_PROVIDERS = [offline_provider_libc_database]
 
-def search_by_hash(hex_encoded_id, hash_type='build_id', unstrip=True):
+
+def search_by_hash(hex_encoded_id, hash_type='build_id', unstrip=True, offline=False):
     assert hash_type in HASHES, hash_type
 
+    # Offline search don't use libcdb cache
```

Comment:
>Why not use the cache?

In the `search_by_symbol_offsets` function, the offline mode doesn't require the use of cache because the purpose of the search is to obtain all matching libc versions, while the cache stores only a subset of them.

For the offline mode in the `search_by_hash` function, cache usage is also omitted because the cache only retains the original libc files. When using `search_by_hash` in the command-line program, it's unable to access information such as `libs_id` and `symbols` from the cache.

>We modify the libc file as well during unstrip, so we'd modify the libc-database base version.

In fact, currently, offline search does not implement the unstrip feature either. This is because the initial envisioned use case for offline functionality is in isolated network environments, and `unstrip_libc` requires an internet connection. Additionally, unstrip relies on using cache, while offline mode does not use cache.